packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -605,6 +610,222 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
+        if (!is(newValue, oldValue)) {
+          changedBits =
+            context.calculateChangedBits !== null
+              ? context.calculateChangedBits(oldValue, newValue)
+              : MAX_SIGNED_32_BIT_INT;
+          if (changedBits !== 0) {
+            propagateContextChange(
+              workInProgress,
+              context,
+              changedBits,
+              renderExpirationTime,
+            );
+          }
+        } else {
+          // No change.
+          changedBits = 0;
+        }
+      }
+      workInProgress.stateNode = changedBits;
+
+      if (oldProps !== null && oldProps.children === newProps.children) {
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
+      const newChildren = newProps.children;
+      reconcileChildren(current, workInProgress, newChildren);
+      return workInProgress.child;
+    } else {
+      return null;
+    }
+  }
+
+  function updateConsumerComponent(
+    current,
+    workInProgress,
+    renderExpirationTime,
+  ) {
+    if (enableNewContextAPI) {
+      const context: ReactContext<any> = workInProgress.type;
+
+      const newProps = workInProgress.pendingProps;
+      const oldProps = workInProgress.memoizedProps;
+
+      // Get the nearest ancestor provider.
+      const providerFiber: Fiber | null = context.currentProvider;
+
+      let newValue;
+      let changedBits;
+      if (providerFiber === null) {
+        // This is a detached consumer (has no provider). Use the default
+        // context value.
+        newValue = context.defaultValue;
+        changedBits = 0;
+      } else {
+        const provider = providerFiber.pendingProps;
+        invariant(
+          provider,
+          'Provider should have pending props. This error is likely caused by ' +
+            'a bug in React. Please file an issue.',
+        );
+        newValue = provider.value;
+        changedBits = providerFiber.stateNode;
+        if (changedBits !== 0) {
+          // Context change propagation stops at matching consumers, for time-
+          // slicing. Continue the propagation here.
+          propagateContextChange(
+            workInProgress,
+            context,
+            changedBits,
+            renderExpirationTime,
+          );
+        }
+      }
+
+      // Store the bits on the fiber's stateNode for quick access.
+      let bits = newProps.bits;
+      if (bits === undefined || bits === null) {
+        // Subscribe to all changes by default
+        bits = MAX_SIGNED_32_BIT_INT;
+      }
+      workInProgress.stateNode = bits;
+
+      if (hasLegacyContextChanged()) {
+        // Normally we can bail out on props equality but if context has changed
+        // we don't do the bailout and we have to reuse existing props instead.
+      } else if (newProps === oldProps && changedBits === 0) {
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
```

Comment:
Why should we ever bail out here? Can't we just let this be handled by the next level? Seems like you'd normally always rerender here if you get here. Can simplify this and avoid the extra working managing and storing previous props.